src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerManager.php

```diff
@@ -70,57 +60,64 @@ public function triggerResourceEvents(array $plugins, array $ids = []): void
     protected function triggerEvents(EventResourceQueryContainerPluginInterface $plugin, array $ids = []): void
     {
         if ($ids) {
-            $this->trigger($plugin, $ids);
+            $this->triggerBulk($plugin, $ids);
 
             return;
         }
 
-        if (!$plugin->queryData($ids)) {
-            $this->trigger($plugin, [static::ID_NULL]);
+        if ($plugin->queryData($ids) === null) {
+            $this->triggerEventWithEmptyId($plugin);
 
             return;
         }
 
-        $this->triggerEventsAll($plugin);
+        $this->processEventsByPluginItreator($plugin);
+    }
+
+    /**
+     * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface $plugin
+     *
+     * @return void
+     */
+    protected function triggerEventWithEmptyId(EventResourceQueryContainerPluginInterface $plugin): void
+    {
+        $this->eventFacade->trigger($plugin->getEventName(), new EventEntityTransfer());
     }
 
     /**
      * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface $plugin
      *
      * @return void
      */
-    protected function triggerEventsAll(EventResourceQueryContainerPluginInterface $plugin): void
+    protected function processEventsByPluginItreator(EventResourceQueryContainerPluginInterface $plugin): void
     {
-        $query = $plugin->queryData();
-        $count = $query->count();
-        $loops = $count / $this->chunkSize;
-        $offset = 0;
-
-        for ($i = 0; $i < $loops; $i++) {
-            $ids = $plugin->queryData()
-                ->offset($offset)
-                ->limit($this->chunkSize)
-                ->where($plugin->getIdColumnName() . ModelCriteria::ISNOTNULL)
-                ->select([$plugin->getIdColumnName()])
-                ->find()
-                ->getData();
-
-            $this->trigger($plugin, $ids);
-            $offset += $this->chunkSize;
+        foreach ($this->createEventResourceQueryContainerPluginIterator($plugin) as $ids) {
+            $this->triggerBulk($plugin, $ids);
         }
     }
 
     /**
-     * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourcePluginInterface $plugin
+     * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface $plugin
+     *
+     * @return \Iterator
+     */
+    protected function createEventResourceQueryContainerPluginIterator(EventResourceQueryContainerPluginInterface $plugin): Iterator
+    {
+        return new EventResourceQueryContainerPluginIterator($plugin, $this->chunkSize);
```

Comment:
This must be cleaned up with an internal factory which is injected from the constructor, @m7moud please create a ticket for this.